fs/fs.go

```diff
@@ -74,20 +74,68 @@ import (
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
 )
 
 var (
 	defaultIndexSelectionPolicy = SelectFirstPolicy
 	fusermountBin               = "fusermount"
+	layerQueueBufferSize        = 1024
 )
 
+type layerQueue struct {
+	queue chan func() string
+	cache *sync.Map
+	smp   *semaphore.Weighted
+}
+
+func newLayerQueue(maxConcurrency int64) *layerQueue {
+	lq := &layerQueue{}
+	lq.queue = make(chan func() string, layerQueueBufferSize) // arbitrarily chosen buffer size
+	lq.cache = &sync.Map{}
+	lq.smp = semaphore.NewWeighted(maxConcurrency)
+	return lq
+}
+
+func (lq *layerQueue) Start(ctx context.Context) error {
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
```

Comment:
Sure, this is a bit simpler.

> I don't think we want an errgroup here

Is this because this is the only one in the group and thus is doing more than what we need?